### ISSUE TITLE
GAN issue

### ISSUE THREAD

**Issue by kushia** [fchollet/keras#4943]
Hi everyone !

I try to train a GAN on cifar10 to generate pretty images. But it seems unable to learn anything.

I tested the discriminator, and he seems to be the problem, because he learns nothing. And I don't see at all from where the problem could come. The loss never decrease or increase (For both the discriminator and generator)

Here is my code : https://github.com/kushia/cifar_gan/blob/master/cifar10_gan.py
I'm really stuck, and it don't seems normal at all to me..
It seems the discriminator is like "Frozen", even when I set the trainable parameter to true..
Thanks you!

**Comment by bstriner** [comment]
Hi @kushia,

I've recently been putting together a lot of GANs so abstracted them out into an `AdversarialModel` class and some utilities. The intent of the module is to make it a little easier to build and train GANs. There are also some examples of how to write callbacks that write image grids, etc. The module allows you to generate the noise on the GPU and have a single function train both D and G.

https://github.com/bstriner/keras_adversarial

Your code actually seems to be training for me. You should have the discriminator train faster than the generator, for example Adam(1e-3) and Adam(1e-4) can work.

Sometimes it converges fastest if you train discriminator then generator on the same noise.

BTW, you should generate noise in one go np.random.uniform(-1, 1, (BATCH_SIZE, 100)), or even better use the GPU.

The model is also way larger than it needs to be for testing purposes. It has 54 million parameters. Start with something small to work out the kinks then scale it up.

Cheers,
Ben

**Comment by kushia** [comment]
Hi @bstriner 
Thanks for you answer!
I'm going to check your repo, it seems very great, thanks!!

Just a little question more, how do you know it's training ?
Cause the loss don't seems to increase or decrease... (I know it should stay more or less at the same level because it's a 'fight', but with gan on mnist, there was some variations)

Thanks for the advices, I'm going to correct that!

**Comment by bstriner** [comment]
It is going to take a long time to hit the same level. The generator only learns anything because the discriminator provides a gradient, so in some sense the discriminator needs to be ahead at all times. While learning, you should see the discriminator loss way below the generator loss.

Interesting blog on GAN stability issues below. They really don't train nicely.

http://www.araya.org/archives/1183

Goodfellow and others have a lot of new tricks which might help.

Cheers,
Ben

**Comment by kushia** [comment]
Thanks for the answer.
But, it is normal than, at the 5th epoch, the loss of discriminator and generator are always very similar ?
discriminator_loss : 0.693169
generator_loss : 0.693344

According to what you say, the discriminator loss should be below... So, there is a problem no ?
And it don't change, it always stay on the same range.

(Thanks for the link, I'll read it attentively )

**Comment by bstriner** [comment]
Another issue: don't have batch normalization before your final sigmoid in the generator. That means the generator can't produce what it wants to no matter what.

**Comment by bstriner** [comment]
Batch normalization for image data should have `axis=1`

**Comment by kushia** [comment]
Previously, I changed the learning rate.
The discriminator loss indeed decrease but after 30 epochs, the results was this.. 
![test](https://cloud.githubusercontent.com/assets/7816846/21734082/e61b91a8-d461-11e6-95ff-a8460da9a07e.png)

Now I deleted the batch normalization before the sigmoid activation and added axis=1 for the others.
Should be ok ?
Thanks again for your help!

**Comment by kushia** [comment]
Oh and, little questions more :
Why the axis should be 1 for batch normalization on image ?
And last, why batch normalization before sigmoid activation would cause the network to fail ?
Thanks!

**Comment by linxihui** [comment]
https://github.com/kushia/cifar_gan/blob/master/cifar10_gan.py#L118
why it is 1 for noise?  Doesn't 1 mean real and 0 mean generated?

**Comment by kushia** [comment]
You mean on this line ?
g_loss = discriminator_on_generator.train_on_batch(noise, [1] * BATCH_SIZE)

This is because, you want to use the gradient error from the discriminator to correct the generator.
(You know it's fake, but the generator need to fool the discriminator)

**Comment by kushia** [comment]
@bstriner 
With the changes for batch norm etc...
And ~10 epochs this is the results : 
![test2](https://cloud.githubusercontent.com/assets/7816846/21734688/f53cf34a-d464-11e6-9e10-c4f7d70e01fd.png)

I have the feeling to do something wrong, but can't figure out what..
Does it seems normal to you ?

**Comment by kushia** [comment]
I was thinking, it's maybe too the way I display generated images with matplotlib ?

**Comment by bstriner** [comment]
Just uploaded a GAN example with a similar architecture but trimmed down. Should be enough to get things going.

https://github.com/bstriner/keras_adversarial/blob/master/examples/example_gan_cifar10.py

After 1 epoch:
![GAN](https://raw.githubusercontent.com/bstriner/keras_adversarial/master/doc/images/gan-cifar10-epoch-000.png)

Probably needs a lot of messing with the architecture to get images that look good but don't take too long to train. If you figure out an architecture that looks better, please let me know and I will update the example.

GANs are weird. If you just keep working with learning rates and complexity you will eventually find something stable. Good luck!

Cheers,
Ben

**Comment by bstriner** [comment]
I just switched to global average pooling in the discriminator instead of flattening. Results are starting to look pretty reasonable if you pull the latest example. Enjoy!

**Comment by kushia** [comment]
That's very impressive!
Great job!
I'm actually looking at your models, and trying to reproduce it..
(But I got some difficulties, because I try to add my touch )
Thanks again for all your help!

**Comment by bstriner** [comment]
Also, I did my testing in theano. I have some function calls that should make it work in tensorflow too, but please let me know if you get something radically different.

GANs are strange beasts. Slightly changing architectures or regularization can get really weird results. They will frequently get stuck into just producing a subsample of images or end up in a cycle that doesn't converge. Just keep trying and best of luck!

Two recent papers on GAN stability and how to deal with them. I actually made `keras_adversarial` so I would have a way to try out unrolled gans in Keras.

https://arxiv.org/abs/1611.02163
https://arxiv.org/abs/1606.03498


Cheers,
Ben

**Comment by kushia** [comment]
Finally, it works!
I used your generator and discriminator, then added some noisy labels (Between 0.7 and 1 for real images and 0. and 0.3 for fake ones)

Works very well!
After 90 epochs, I have really good results.
The networks invents some animals clearly, but that's really funny!

And you're right, as I saw, it's very hard to stabilize the GAN.
I would never have it without you, a big thanks!
Do we have an idea or intuition of why GAN are so sensitive to learning rate and regularization ?

Thanks for the papers, going to read that!

Have a nice day!

**Comment by bstriner** [comment]
I can give you my theories on GANs but it's a hot research topic right now. No one really knows the right answer yet but a lot of people are trying. Those papers and the blog should be a good place to start.

One of the intentions behind `keras_adversarial` is that you can use it just to train adversarial models. If you want to do more, subclass `AdversarialOptimizer` and try to find better strategies. The training strategy is separated from the modelling.

Here is what the current CIFAR10 example is looking like at epoch 99. Please let me know if you find something that looks better and I will update the example. This version uses 1.8M generator parameters and 1M discriminator parameters with dropout. Takes about 4 minutes per epoch.

![CIFAR10](https://github.com/bstriner/keras_adversarial/raw/master/doc/images/gan-cifar10-epoch-099.png)

https://github.com/bstriner/keras_adversarial/blob/master/examples/example_gan_cifar10.py

Cheers,
Ben

**Comment by kushia** [comment]
Your results are very impressive!
I got some good results too, but yours are a lot better!

Thanks again, I will use your keras_adversarial now!
I'll warn you if I can find a better model. GAN are really very funny!
And thank you so much for your help!

Jean

**Comment by bstriner** [comment]
If there's something worth adding, just add an issue to the repo. Best of luck!

**Comment by sixtinedpdc** [comment]
Hi everyone, 

I tried to train this gan example: https://github.com/bstriner/keras_adversarial/blob/master/examples/example_gan_cifar10.py. But I don't understand one thing. How do you obtain the pictures produced by the generator? I wanted to use the method "on_epoch_end" to record the images, but I could register only one set of images. I was thinking about register images regularly, using an interval, but I didn't manage to do that.
Do you have an idea to do that? 
Thank you!

**Comment by bstriner** [comment]
@sixtinedpdc Check out the code for `ImageGridCallback` and `write_image_grid`. What specifically are you trying to implement? I provided simple callbacks for writing images. Should be easy enough to edit the callbacks for your own purposes. If you add something useful let me know or make a PR and I might bring it in to master.

https://github.com/bstriner/keras-adversarial/blob/master/keras_adversarial/image_grid_callback.py
https://github.com/bstriner/keras-adversarial/blob/master/keras_adversarial/image_grid.py

`ImageGridCallback` runs at the end of each epoch. It runs a function handle and writes the output of that function handle to an image file. The user supplies a function to the callback that generates a numpy array each epoch.

That callback lambda happens to be named `generator` but it is just any function that can be run periodically to generate a numpy array of images. The `ImageGridCallback` takes care of creating directories and writing image files, but if you want to write something else you can adapt the code.

The `ImageGridCallback` is a standard Keras callback you can read about elsewhere. 
* If you want to run more than once an epoch, override `on_batch_end` instead of `on_epoch_end`. Write images if `batch%100==0` or something like that.
* If you want less than once an epoch, change `on_epoch_end` to modulus something.
* If you have several different images you want to generate, add multiple `ImageGridCallback` objects. The AAE and BiGAN examples use 2 callbacks, one for autencoded images and one for generated images.

Make sure to generate a unique path for each image whatever changes you make so you don't overwrite yourself.

In the examples, I make a function with no parameters that samples from the latent space, runs the data through keras, and rehapes the output to look like an array of images. I create an `ImageGridCallback` using that function. (There is a lot of other stuff for compatibility between TF and theano that might make reading it harder, so this is just the core)

```python
zsamples = np.random.normal(size=(10 * 10, latent_dim))
 def generator_sampler():
        xpred = generator.predict(zsamples)
        return xpred.reshape((10, 10) + xpred.shape[1:])
generator_cb = ImageGridCallback(os.path.join(path, "epoch-{:03d}.png"), generator_sampler, cmap=None)
```

Depending on your architecture, but especially with GANs and language modelling, your prediction model for testing is going to be different than your training model.

In the GAN examples, you train on a combined discriminator and generator model, but the images are output by just the generator submodel.

Cheers

**Comment by sixtinedpdc** [comment]
Hi @bstriner ,

Thanks a lot for you answer ! It helped me a lot because I didn't understand the role of the callback class. Now it's clearer. I will try to do what you have said to produce more images (not only at the end of an epoch). I tell you when I have good results!

Cheers
